Добавляет смену движения автомобиля при столкновении и конца карты, и вычисляет урон от столкновения

maver1cus opened the pull request.

maver1cus committed:
refactor: переименовывает метод класса Car

maver1cus committed:
feat: изменяет метод getDirectionRide класса Car
теперь возвращает текущее directionRide car

maver1cus committed:
feat: реализует смену направляния движения Car при столкновении

maver1cus committed:
feat: реализует emitter и нанесение урона при столконовении

kicumkicum commented on src/const.js:
```diff
@@ -9,3 +9,5 @@ export const SymbolsForPrintElements = {
   ITEM: `+`
 };
+
+export const countDamageToCrash = 10;
```
Константы стоит именовать `ВОТ_ТАКИМ_СТИЛЕМ` - `COUNT_DAMAGE_TO_CRASH`

kicumkicum replied:
И тут лучше подходит слово `value`, вместо `count`

maver1cus replied:
done

kicumkicum commented on src/engine/car.js:
```diff
@@ -1,24 +1,57 @@
+      this._emitter.emit('car:destroy', this);
+    }
+    this._changeDirection();
+  }
```
Добавь в эту функцию переносов строк, что бы было проще читать

maver1cus replied:
done

kicumkicum commented on src/engine/car.js:
```diff
@@ -1,24 +1,57 @@
     this._life = life;
     this._speed = speed;
     this._directionRide = directionRide;
+    this._emitter = emitter;
```
`emitter` слишком общее название. Не очень понятно, что за ним скрывается, кроме того, что он шлет ивенты.

maver1cus replied:
done

kicumkicum commented on src/engine/car.js:
```diff
@@ -1,24 +1,57 @@
+    this._emitter.subscribe('world:end', this._worldEnd.bind(this));
+  }
+
+  _crash(element) {
```
Так как это обработчик события, то его стоит именовать с помощью суффиксов `on` или `handle`

maver1cus replied:
done

kicumkicum commented on src/engine/emitter.js:
```diff
@@ -0,0 +1,24 @@
+    this.listeners = {};
+  }
+
+  emit(event, ...args) {
```
О, ты сам его написал. Офигенно!

kicumkicum replied:
Но добавь переносов внутри методов

maver1cus replied:
done

kicumkicum commented on src/engine/world.js:
```diff
@@ -23,29 +26,42 @@ export default class World {
-        coords = this._validateCoords(coords);
-        this._elements.set(element, coords);
+        if (!this._isValidCoords(candidateCoords)) {
+          this._emitter.emit('world:end', element);
```
Названия событий стоит помещать в константы класса, который их испускает. Так проще работать с ними и отслеживать зависимости одних сущностей от других.

maver1cus replied:
done

kicumkicum reviewed (changes requested):
Серёжа, все офигенно! Молодец 👍🏽

maver1cus committed:
refactor: переименовывает классы и их методы

kicumkicum commented on index.js:
```diff
@@ -1,9 +1,9 @@
+import BusEvents from './src/engine/bus-events.js';
 
-const emitter = new Emitter();
+const emitter = new BusEvents();
```
А вот тут наоборот. Класс - это чистый event-emitter без конкретики. Но инстанс - это уже шина событий - появляется конкретика. Поэтому класс не нужно было переименовывать, а инстанс и свойство в `Car` нужно

maver1cus replied:
done

kicumkicum commented on src/engine/world.js:
```diff
@@ -1,4 +1,5 @@
 import {Directions} from '../const.js';
+import Car from './car.js';
```
Смотри, у тебя появилась циклическая зависимость. Car зависит от World, а World от Car. Шина событий должна была развязать это, но пока не вышло

maver1cus replied:
done

kicumkicum reviewed (changes requested):
Раньше циклическая зависимость была не явна. Сейчас она стала явная. Давай попробуем вообще от нее уйти

maver1cus committed:
refactor: правки
- делает Car наследником Item и переносит туда общие методы
- переносит события Emitter в BusEvent
- изменяет нанесение урона при столкновении

kicumkicum commented on src/engine/item.js:
```diff
@@ -1,5 +1,37 @@
+    }
+
+    element._life = element._life - markElement._valueDamageToCrash;
+    markElement._life = markElement._life - element._valueDamageToCrash;
```
Тут ты обращаешься к приватным свойствам объекта снаружи. А так делать нельзя

maver1cus replied:
done

maver1cus committed:
Removes access to private properties of Item from other entities
